File: 04-herencia/juego_de_estrategia/soldado.py
"""Class Soldado"""

from unidad import Unidad


class Soldado(Unidad):
    def __init__(self, posicion):
        super().__init__(200, 10, posicion)
        self.__energia = 100

    @property
    def energia(self):
        return self.__energia

    def atacar(self, enemigo: Unidad):
        if self.puede_atacar(enemigo):
            self.__energia -= 10
            enemigo.recibir_danio(super().danio)

    def puede_atacar(self, enemigo: Unidad) -> bool:
        return (
            super().distancia(enemigo) == 0
            and not enemigo.esta_muerto()
            and self.energia >= 10
        )

File: workspace-2025-1c/Herencia/src/juegoDeEstrategia/Soldado.java
package juegoDeEstrategia;

public class Soldado extends Unidad implements Bebedor {

	private int energia = 100;

	public Soldado(int posicion) {
		super(200, 10, posicion);
	}

	public int getEnergia() {
		return this.energia;
	}

	@Override
	public void atacar(Unidad enemigo) {

		if (this.puedeAtacar(enemigo)) {
			this.energia -= 10;
			enemigo.recibirDanio(10);
		}
	}

	public boolean puedeAtacar(Unidad enemigo) {
		if (enemigo.estaMuerto()) {
			throw new Error("No podes atacar a un muerto");
		}

		return this.distancia(enemigo) == 0 && !this.estaMuerto() && this.energia >= 10;
	}

	@Override
	public void beberAgua() {
		this.energia = 100;
	}

}
